### PR TITLE
ui: add button labels that tell the user which button does what

### PR DESCRIPTION
**Pull request by hnez**
This adds a little legend on the right of the screen (where the two buttons are) that tells the user what which button does what:

dig_out
iobus
iobus_health
locator
overtemperature
power
power_fail
reboot
screensaver
system
uart
update_available
update_installing
usb
usb_overload
reboot_slot

TODO before merging:

  - [x] Add comments to the drawing routine that make it easier to read
  - [x] Add comments to the screen rotation trickery
  - [x] Add legends to alert screens as well

### CODE REVIEW

File: src/ui/display.rs [hunk 1]
```diff
@@ -176,3 +186,28 @@ impl OriginDimensions for DisplayExclusive {
...
+
+impl<'a> DrawTarget for DisplayRotated<'a> {
+    type Color = BinaryColor;
+    type Error = core::convert::Infallible;
+
+    fn draw_iter<I>(&mut self, pixels: I) -> Result<(), Self::Error>
+    where
+        I: IntoIterator<Item = Pixel<Self::Color>>,
+    {
+        let res = self.size();
```
Comment (hnez) [inline src/ui/display.rs:198]: Add a comment about this only doing one kind of rotation (the one we need for the button legend).

File: src/ui/display.rs [hunk 2]
```diff
@@ -176,3 +186,28 @@ impl OriginDimensions for DisplayExclusive {
...
+            .into_iter()
+            .map(|Pixel(coord, color)| Pixel(Point::new(coord.y, res_y - coord.x), color));
+
+        self.inner.draw_iter(rotated_pixels)
+    }
+}
+
+impl<'a> OriginDimensions for DisplayRotated<'a> {
+    fn size(&self) -> Size {
+        self.inner.size()
```
Comment (hnez) [inline src/ui/display.rs:211]: Our display is square, but this should flip the x/y coordinates anyways.

File: src/ui/screens/system.rs [hunk 3]
```diff
@@ -93,8 +95,8 @@ impl ActivatableScreen for SystemScreen {
                 display,
                 row_anchor(1),
                 Box::new(|info: &LinkInfo| match info.carrier {
-                    true => format!("Uplink: {}MBit/s", info.speed),
-                    false => "Uplink: Down".to_string(),
+                    true => format!("UL:  {}MBit/s", info.speed),
+                    false => "UL:  Down".to_string(),
```
Comment (hnez) [inline src/ui/screens/system.rs:99]: Split this into a separate commit and write about the _why_ in the commit message (the IP address line got too wide with the button legend so it had to move left without breaking alignment of the rows, so "Uplink" had to become shorter).

File: src/ui/screens/system.rs [hunk 4]
```diff
@@ -75,6 +75,8 @@ impl ActivatableScreen for SystemScreen {
     fn activate(&mut self, ui: &Ui, display: Display) -> Box<dyn ActiveScreen> {
         draw_border("System Status", SCREEN_TYPE, &display);
 
+        display.with_lock(|target| draw_button_legend(target, "Next / Do", "Screen"));
+
```
Comment (hnez) [inline src/ui/screens/system.rs:79]: `draw_border()` followed by `with_lock(|t| draw_button_legend)` seems to be a common pattern now. Check if we should make `draw_border` accept a `DisplayExclusive` like `draw_button_legend` does, so both can be drawn in the same `.with_lock()` call.

File: src/ui/widgets.rs [hunk 5]
```diff
@@ -98,6 +99,51 @@ pub struct DynamicWidget<T: Sync + Send + 'static> {
...
+
+    Line::new(Point::new(26, 225), Point::new(26, 240))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
+
+    Line::new(Point::new(214, 225), Point::new(214, 240))
+        .into_styled(PrimitiveStyle::with_stroke(BinaryColor::On, 1))
+        .draw(&mut target)
+        .unwrap();
```
Comment (hnez) [inline src/ui/widgets.rs:140]: Add a comment for each draw call.

File: src/ui/widgets.rs [hunk 6]
```diff
@@ -248,6 +294,22 @@ impl<T: Serialize + DeserializeOwned + Send + Sync + Clone + 'static> DynamicWid
         )
     }
 
+    pub fn button_legend(
```
Comment (hnez) [inline src/ui/widgets.rs:297]: Add a docstring.

**Comment by hnez** [comment]
I've now gone through all screens and added legends.
With that I am ready for review.

**Comment by SmithChart** [comment]
First: I find this really helpful. It feels like the first time that I really know what button does which action. Or at least this way a user can learn to use them right. (Maybe we can add caps with different colors to the buttons, to make it even easier to remember?)

@fiurin  and I had a look at this change from a user-perspective. One thing we kind of stumbled across is the label on the lower button. This button can either be used with a short- or long press. And does different things depending on the type of press.
For screens with more than one option the `toggle` or `select` labels are not clear. We would vote to use `Action` here instead. This is intended to be ominous and vague, so the user can assume that both long- and short press are available here. This applies to the "USB Host", "Digital out", "DUT UART" and "System screens".

I would vote to leave all other screens as they are in this PR.

**Comment by hnez** [comment]
I've rebased the PR on top of recent main (it should go in after #59 though, so we see the CI succeed before merging) and changed the labels in the mentioned screens from what they were there to "Action".

**Comment by SmithChart** [comment]
> I've rebased the PR on top of recent main (it should go in after #59 though, so we see the CI succeed before merging) and changed the labels in the mentioned screens from what they were there to "Action".

Thanks! I tried the recent changes on my device. Looks good!

**Comment by hnez** [comment]
Using an actual graphics toolkit, like LVGL for instance, would be great I think, because it could make the UI more modern (and maybe make working on the UI a bit easier).
I will propose an evaluation of GUI options for the TAC and other products with tiny displays and reduced input options as a techweek project once that comes around.
